packages/jupyter-ai/src/hooks/use-replace.ts

```diff
@@ -0,0 +1,56 @@
+import { useActiveCellContext } from '../contexts/active-cell-context';
+import { useSelectionContext } from '../contexts/selection-context';
+
+export type UseReplaceReturn = {
+  /**
+   * If a user has a range of text selected, this function replaces the
+   * selection range with `value`. Otherwise, if the user has an active notebook
+   * cell, this function replaces the contents of the active cell with `value`.
+   *
+   * Otherwise (if a user does not have a text selection or active cell), this
+   * function does nothing.
+   */
+  replace: (value: string) => unknown;
+  /**
+   * Whether the replace button should be disabled, i.e. the user does not have
+   * a text selection or active cell.
+   */
+  replaceDisabled: boolean;
+  /**
+   * Label that should be shown by the replace button using this hook.
+   */
+  replaceLabel: string;
+};
+
+/**
+ * Hook that provides a function to either replace a text selection or an active
+ * cell. Manages related UI state. Should be used by any button that intends to
+ * replace some user selection.
+ */
+export function useReplace(): UseReplaceReturn {
+  const [textSelection, replaceTextSelection] = useSelectionContext();
+  const activeCell = useActiveCellContext();
+
+  const replace = (value: string) => {
+    if (textSelection) {
+      replaceTextSelection({ ...textSelection, text: value });
+    } else if (activeCell.exists) {
+      activeCell.manager.replace(value);
+    }
+  };
+
+  const replaceDisabled = !(textSelection || activeCell.exists);
+
+  const numLines = textSelection?.text.split('\n').length || 0;
+  const replaceLabel = textSelection
```

Comment:
This may be causing the selection of an entire line even if it not highlighted in the text selection. See the comment and video here: https://github.com/jupyterlab/jupyter-ai/pull/895#issuecomment-2226432797. Or it is coming from `

Reply:
This is because when the cursor is dragged like that, it is including the `\n` character after the last selected letter. This causes the line after the selection range to be appended directly to the last line of the replacement text. See this video, which shows this by scrolling to the right:

https://github.com/user-attachments/assets/93793a12-0fa7-4a91-8e8a-215db69926b4

I think this behavior should remain as-is to preserve parity with the CodeMirror editor. The same thing will happen if you select the same range and type a character:

https://github.com/user-attachments/assets/c64e595f-08de-4809-9bb4-3feae9106130